Fix cli tests in windows

ahtrotta opened the pull request:
Fixes getappmap/board#188

dividedmind commented on packages/cli/tests/sleeper.js:
```diff
@@ -0,0 +1,5 @@
+const timeout = process.argv[2] || 1000;
+
+(async function sleep() {
+  await new Promise((resolve) => setTimeout(resolve, timeout));
+})();
```
Plain `setTimeout(() => {}, timeout)` at the top level should work too.

dividedmind commented on packages/cli/tests/unit/agentInstall/javaBuildToolInstaller.spec.ts:
```diff
@@ -16,20 +16,21 @@ describe('JavaBuildToolInstaller', () => {
 
   describe('when IntelliJ is installed', () => {
-    const intellijHome = '/IntelliJ';
+    const intellijHome = (process.platform === 'win32' ? '\\' : '/') + 'IntelliJ';
```
You could also use `path.sep` or, even clearer, `path.join('/', 'IntelliJ')` (I checked, it does the right thing on windows.)

ahtrotta replied:
Yeah, that's cleaner. I changed it.

dividedmind commented on packages/cli/tests/unit/agentInstall/javaBuildToolInstaller.spec.ts:
```diff
@@ -16,20 +16,21 @@ describe('JavaBuildToolInstaller', () => {
     it('appends JetBrains JDK to PATH', () => {
       addJetBrainsEnv();
-      expect(process.env.PATH).toMatch(/\/IntelliJ\/jbr\/bin/);
+      expect(process.env.PATH).toMatch(/[\/\\]IntelliJ[\/\\]jbr[\/\\]bin/);
```
`toContain(path.join(intellijHome, 'jbr', 'bin'))` would also verify the right separator was used.

dividedmind commented on packages/cli/tests/unit/agentInstall/javaBuildToolInstaller.spec.ts:
```diff
@@ -16,20 +16,21 @@ describe('JavaBuildToolInstaller', () => {
       delete process.env.JAVA_HOME;
       addJetBrainsEnv();
-      expect(process.env.JAVA_HOME).toBe(intellijHome + '/jbr');
+      const expectedIntellijHome = intellijHome + (process.platform === 'win32' ? '\\jbr' : '/jbr');
```
This wants to be `path.join` :)

ahtrotta replied:
True!

dividedmind commented on packages/cli/tests/unit/depends.spec.ts:
```diff
@@ -10,7 +10,7 @@ import { listAppMapFiles, verbose } from '../../src/utils';
 
 const fixtureDir = join(__dirname, 'fixtures', 'ruby');
-const now = Date.now();
+const now = Date.now() / 1000;
```
I don't understand this change.

ahtrotta replied:
[Apparently, `utimesSync` takes timestamps in seconds, but `Date.now()` generates them in milliseconds](https://github.com/nodejs/node/issues/33227#issuecomment-623391936). But I'm confused as to why it works in Linux / Mac if that's the case...

ahtrotta replied:
I agree, I implemented this

dividedmind commented on packages/cli/tests/unit/depends.spec.ts:
```diff
@@ -20,6 +20,7 @@ describe('Depends', () => {
   beforeAll(async () => verbose(process.env.DEBUG === 'true'));
   beforeEach(async () => {
     appMapDir = tmp.dirSync().name;
+    expected = join(appMapDir, 'user_page_scenario').split('\\').join('/');
```
`replaceAll` could work better :) also, if you just want to ensure usage of unix separators, you can use `path.posix.join`.

ahtrotta replied:
For some reason, when I use `replaceAll` I get this error:
![image](https://user-images.githubusercontent.com/45714532/195427239-1323abbb-7cd3-4469-a767-1b87449fbc65.png)

I changed it to `replace`, though.

dividedmind commented on packages/cli/tests/unit/encodedFile.spec.ts:
```diff
@@ -55,7 +55,8 @@ describe('EncodedFile', () => {
   it('writes a 7-bit ASCII file', () => {
-    verifyUpdate('ascii', '\n');
+    const newlineChar = process.platform === 'win32' ? '\r\n' : '\n';
+    verifyUpdate('ascii', newlineChar);
```
I don't understand how it now passes on Windows. The test file uses `\n`, so the test should now fail, right?

BTW, perhaps it would be a good idea to move those ascii tests into the per-platform section above.

ahtrotta replied:
I'm not sure, but I've noticed that all files have `\r\n` by default in Windows. Even when I'm pulling them from GitHub, where they were generated with `\n`.

When I move the test into the per-platform section, I get these errors:
![image](https://user-images.githubusercontent.com/45714532/195431018-a0b0b326-24c3-48b5-b0d7-2e8ed292e1cb.png)

dividedmind replied:
Yup, that's probably the problem of your git configuration (autocrlf).

dividedmind commented on packages/cli/tests/unit/inspect.spec.ts:
```diff
@@ -11,9 +11,9 @@ import { CodeObject, CodeObjectMatch } from '../../src/search/types';
 
 const fixtureDir = join(__dirname, 'fixtures', 'ruby');
-const appMapDir = tmp.dirSync().name;
+const appMapDir = tmp.dirSync().name.split('\\').join('/');
```
Or `replaceAll` ;)

dividedmind replied:
It is weird. Too old a typescript or some type definitions? But I'd think they'd be pinned by yarn.

dividedmind commented on packages/cli/tests/unit/inspect.spec.ts:
```diff
@@ -11,9 +11,9 @@ import { CodeObject, CodeObjectMatch } from '../../src/search/types';
+const appMapDir = tmp.dirSync().name.split('\\').join('/');
 
-const now = Date.now();
+const now = Date.now() / 1000;
```
This again..?

ahtrotta replied:
[Apparently, `utimesSync` takes timestamps in seconds, but `Date.now()` generates them in milliseconds](https://github.com/nodejs/node/issues/33227#issuecomment-623391936). But I'm confused as to why it works in Linux / Mac if that's the case...

dividedmind commented on packages/cli/tests/unit/record/testCaseRecording.spec.ts:
```diff
@@ -51,12 +55,24 @@ describe('record.testCaseRecording', () => {
+        '/bin/sh -c sleep 1';
+
+      // windows will throw an error when a process is killed using the pid
+      const windowsError = '\nTest command failed with status code 1: C:\\WINDOWS\\system32\\cmd.exe /d /s /c "node .\\tests\\sleeper.js 1000"';
```
This constant could be moved down to the `if (process.platform)`; also it could use `fullExpectedCommand` to make it a little DRYer.

ahtrotta replied:
Good idea!

dividedmind reviewed (changes requested):
Looks good, thanks! I had some questions, and some stylistic suggestions but you're free to take them or leave them.

The only thing I'm particularly (not too much though) concerned about is the prettier change. I'd rather we don't see crlf in the repository one day ;)

If the tests are fixed, can we change travis configuration to run the full CLI test suite on Windows now?

ahtrotta commented on packages/cli/tests/unit/agentInstall/installCommand.spec.ts:
```diff
@@ -28,7 +28,7 @@ import { withStubbedTelemetry } from '../../helper';
 tmp.setGracefulCleanup();
 
-const invokeCommand = (
+const invokeCommand = async (
```
True, I changed it back :)

ahtrotta commented on packages/cli/tests/unit/agentInstall/installCommand.spec.ts:
```diff
@@ -1053,7 +1053,7 @@ packages:
 
   describe('Multi-project install flow', () => {
-    let expectedStubs;
+    let expectedStubs: any[];
```
Okay, good to know, I changed it to `SinonStub[]`

ahtrotta commented on packages/cli/tests/unit/record/action/configureHostAndPort.spec.ts:
```diff
@@ -11,11 +11,12 @@ describe('record.action.configureHostAndPort', () => {
     context = new RecordContext(config);
-    return context.initialize();
+    await context.initialize();
+    return;
```
True, changed it back

ahtrotta commented:
Thanks for the excellent review! I addressed all of your comments. Let me know if you have any other questions or concerns.

I made it so all of the tests except for one run in Windows on Travis.

dividedmind commented on .prettierrc.cjs:
```diff
@@ -4,7 +4,7 @@ module.exports = {
   singleQuote: true,
   arrowParens: 'always',
-  endOfLine: 'lf',
+  endOfLine: 'auto',
```
This is weird. Sounds like something is changing your line endings; I suspect git. Maybe you need to change your git config? `git config --global core.autocrlf input` or `git config --global core.autocrlf false` should do the trick.

dividedmind commented on packages/cli/tests/testFilter.js:
```diff
@@ -0,0 +1,17 @@
+const FILTERED_TESTS = process.platform === 'win32' ? ['fingerprintWatchCommand'] : [];
```
Why does this test need to be filtered? If there's something left to fix for this issue, that's fine, but please remember to create another one to keep track of the loose end.

ahtrotta replied:
This one gives this error in Travis on Windows:

![image](https://user-images.githubusercontent.com/45714532/195657050-9c69070d-3124-402d-a239-23b6bc1efc9e.png)

You can see it in the logs on this build: https://app.travis-ci.com/github/getappmap/appmap-js/jobs/585469980#L220

I'll make a new issue for it.

dividedmind commented:
> Thanks for the excellent review! I addressed all of your comments. Let me know if you have any other questions or concerns.
> 
> I made it so all of the tests except for one run in Windows on Travis.

Thanks, that looks much better! Now I'm only worried about those line endings. Perhaps it would be a good idea to add a `gitattributes` per-file eol override for the affected expectation files to prevent git from breaking the tests. (It'd probably look like ```
packages/cli/tests/unit/fixtures/encodedFile/**/* eol=lf
packages/cli/tests/unit/fixtures/encodedFile/windows/* eol=crlf
``` — haven't checked if it actually works, YMMV. Another option would be to do `-text` for all these files so git leaves them alone.)

ahtrotta commented:
I put all of the tests into the platform-specific `describe` blocks in `encodedFile.spec.ts`, and all of the tests except for one are passing: 6ea720473bb0c5c934156f0471a26ce9cfdf478c

The strange thing is that this test passes for me locally on Windows and on Linux, so I'm not sure how to debug this. I tried changing the `.travis.yml` file, but it didn't do anything.

Do you have any thoughts?

ahtrotta commented:
Okay, I got it to work. Turns out that forcing `CRLF` in the Windows files using:
```
/packages/cli/tests/unit/fixtures/encodedFile/windows/**/* eol=crlf
```
was causing issues on Travis. I'm not sure why. But I needed to force `LF` for the Unix test files for it to work on Windows.

Let me know if there's anything else you think needs to be changed!

dividedmind reviewed (approved):
Looks great now, thanks for figuring out this pesky EOL business!

(Please remember to rebase and fixup everything before merging.)

appland-release commented:
:tada: This PR is included in version @appland/appmap-v3.45.0 :tada:

The release is available on:
- `@appland/appmap-v3.45.0`
- [GitHub release](https://github.com/getappmap/appmap-js/releases/tag/%40appland/appmap-v3.45.0)
- `NPM publish`

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
